core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
-            if !self.is_populated()? {
-                return Ok(true);
-            }
+    /// Kills all processes in this cgroup
+    pub fn kill(&self) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
 
-            // Check again if time is exceeded
-            leftover_time = timeout.saturating_sub(start.elapsed());
-            if leftover_time <= Duration::ZERO {
-                return Ok(false);
-            };
-
-            // Poll wait for event with given leftover time as timeout
-            poller
-                .wait(Vec::new().as_mut(), Some(leftover_time))
-                .typ(SystemError::Panic)?;
-            poller
-                .modify(event_file.as_raw_fd(), Event::readable(42))
-                .map_err(anyhow::Error::from)
-                .typ(SystemError::Panic)?;
-
-            // determine new leftover time
-            leftover_time = timeout.saturating_sub(start.elapsed());
+        // We need to check for the existance of cgroup.kill, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.kill");
+        if !path.exists() {
+            bail!("cannot kill the root cgroup");
         }
+
+        Ok(fs::write(path, "1")?)
     }
 
-    /// Deletes a cgroup by killing it's processes and removing the directory afterwards
-    pub fn delete(&self, timeout: Duration) -> TypedResult<()> {
-        self.kill_all_timeout(timeout)?;
-        for d in WalkDir::new(&self.path)
-            .into_iter()
-            .flatten()
-            .filter(|e| e.file_type().is_dir())
-            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
-        {
-            std::fs::remove_dir(&d.path()).typ(SystemError::CGroup)?;
-            trace!("Removed {:?}", d.path().as_os_str())
+    /// Returns the path of this cgroup
+    pub fn get_path(&self) -> PathBuf {
+        self.path.clone()
+    }
+
+    /// Returns the path of the event file, which may be polled
+    pub fn get_events_path(&self) -> PathBuf {
+        self.path.join("cgroup.events")
+    }
+
+    /// Kills all processes and removes the current cgroup
+    pub fn rm(&self) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!(
+                "{} is not a valid cgroup",
+                self.path.to_str().unwrap_or("N/A")
+            );
+        }
+
+        // Recursively delete all sub cgroups
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+
+            if meta.is_dir() {
+                // Each sub cgroup is deleted by calling this function recursively
+                let child = Self::import_root(&entry.path())?;
+                child.rm()?;
+            }
         }
+
+        // Remove the actual cgroup
+        self.kill()?;
+        while self.populated()? {}
```

Comment:
Isnt this very CPU intensive?

Reply:
Previously the event file was polled with `poller::Poller`

Reply:
I think that a poller isn't worth the complexity in this case. Sending `SIGKILL` to many processes is a very cheap task. Killing all processes on my system (shutdown) takes less than 500ms and this is the worst case, because it contains ALL processes that the system has. Maybe it could become more efficient by adding something that is known as a `noop` in Assembly. However, I do not know if and how Rust supports this and inline-assembly is an absolute no-go.

Reply:
I have very good news! I have looked at the kernel source code and although I barely know anything about kernel internals, it looks like -- at least on the first view --, that `write(2)` calls to the killfile only return, after all processes have already been killed, making the polling approach useless.